packages/flutter/lib/src/widgets/overlay.dart

```diff
@@ -1723,6 +1727,14 @@ class OverlayPortalController {
 /// widget rebuilds. Stateful descendants in the overlay child subtree may lose
 /// states as a result.
 ///
+/// {@template flutter.widgets.overlayPortal.targetsOverlay}
+/// Use [targetsOverlay] to choose which [Overlay] this overlay portal
+/// attaches to. Consider using [GlobalKey] or [Overlay.of] with appropriate
+/// [BuildContext] to get the [OverlayState]. One can also use
+/// `Overlay.of(context, rootOverlay: true)` to pick the root [Overlay]. If
```

Comment:
Hmm calling `Overlay.of` doesn't establish a dependency (it's not using an inheritedWidget). So if an OverlayPortal gets reparented to a different subtree where it can't see the old target Overlay, it's not guaranteed that the OverlayPortal's parent will rebuild to update the `targetsOverlay` parameter (e.g., if you cache the widget).

Reply:
See: https://main-api.flutter.dev/flutter/widgets/Overlay/maybeOf.html

Reply:
(I thought I had a test for this but apparently all tests are passing)

Reply:
In that case, it is the consumer of this widget to ensure it grabs the correct overlay state when it change? (in State.activate) not entirely sure what else we can do. using globalkey is not feasible unless we expose root overlay's globalkey all the way to WidgetsApp

Reply:
>  it is the consumer of this widget to ensure it grabs the correct overlay state when it change

Yeah if the creator of the OverlayPortal widget rebuilds. But the problem is what if that widget doesn't rebuild because the user cached the widget? I think we need an alternative version of `Overlay.of` that establishes a dependency (i.e. an InheritedWidget for `Overlay`). To prevent these kind of bugs maybe the new `of` method should use a different type from `OverlayState` so people can't use `Overlay.of` in these cases.

Reply:
> Can the overlay.of create dependency be another pr?

Yeah but I think the `Overlay.of` that establishes dependencies a prerequisite for this PR?

Reply:
> that both the old parent Element and new parent Element rebuild)?

If you don't have dependecy, but have some specialize logic to update state in activate, then this makes sense. My guess is that since State has access to to activate callback, this is necessary to make it not a footgun.

> Yeah but I think the Overlay.of that establishes dependencies a prerequisite for this PR?

sgtm